drivers/flash/flash_realtek_rts5912.c

```diff
@@ -0,0 +1,836 @@
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	spic_reg->FLUSH = SPIC_FLUSH_ALL_Msk;
+}
+
+static void spic_cs_active(void)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	spic_reg->SER = 1;
+}
+
+static void spic_cs_deactivate(void)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	spic_reg->SER = 0ul;
+}
+
+static void spic_usermode(void)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	spic_reg->CTRL0 |= SPIC_CTRL0_USERMD_Msk;
+}
+
+static void spic_automode(void)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	spic_reg->CTRL0 &= ~SPIC_CTRL0_USERMD_Msk;
+}
+
+static SPIC_STATUS_t spic_prepare_command(const SPIC_COMMAND_t *command, uint32_t tx_size,
+					  uint32_t rx_size, uint8_t write)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	uint32_t i;
+	uint8_t addr_len = user_addr_len[command->address.size];
+
+	spic_flush_fifo();
+
+	/* set SSIENR: deactivate to program this transfer */
+	spic_reg->SSIENR = 0ul;
+
+	/* set CTRLR0: TX mode and channel */
+	spic_reg->CTRL0 &= ~(TMOD(3) | CMD_CH(3) | ADDR_CH(3) | DATA_CH(3));
+	spic_reg->CTRL0 |= TMOD(write == 0x01 ? 0x00ul : 0x03ul) |
+			   ADDR_CH(command->address.bus_width) | DATA_CH(command->data.bus_width);
+
+	/* set USER_LENGTH */
+	spic_reg->USERLENGTH = USER_CMD_LENGTH(1) |
+			       USER_ADDR_LENGTH(command->address.disabled ? 0 : addr_len) |
+			       USER_RD_DUMMY_LENGTH(command->dummy_count * spic_reg->BAUDR * 2);
+
+	/* Write command */
+	if (!command->instruction.disabled) {
+		spic_reg->DR.BYTE = command->instruction.value;
+	}
+
+	/* Write address */
+	if (!command->address.disabled) {
+		for (i = 0; i < addr_len; i++) {
+			spic_reg->DR.BYTE =
+				(uint8_t)(command->address.value >> (8 * (addr_len - i - 1)));
+		}
+	}
+
+	/* Set TX_NDF: frame number of Tx data */
+	spic_reg->TXNDF = TX_NDF(tx_size);
+
+	/* Set RX_NDF: frame number of receiving data. */
+	spic_reg->RXNDF = RX_NDF(rx_size);
+
+	return SPIC_STATUS_OK;
+}
+
+static SPIC_STATUS_t spic_transmit_data(const void *data, uint32_t *length)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	uint32_t i, len = *length;
+
+	/* set SSIENR to start the transfer */
+	spic_reg->SSIENR = SPIC_SSIENR_SPICEN_Msk;
+
+	/* write the remaining data into fifo */
+	for (i = 0; i < len;) {
+		if (spic_reg->SR & SPIC_SR_TFNF_Msk) {
+			spic_reg->DR.BYTE = ((uint8_t *)data)[i];
+			i++;
+		}
+	}
+	return SPIC_STATUS_OK;
+}
+
+static SPIC_STATUS_t spic_receive_data(void *data, uint32_t *length)
+{
+	SPIC_Type *spic_reg = SPIC_REG_BASE;
+
+	uint32_t i, cnt, rx_num, fifo, len;
+	uint8_t *rx_data = data;
+
+	len = *length;
+	rx_data = data;
+
+	/* set SSIENR to start the transfer */
+	spic_reg->SSIENR = SPIC_SSIENR_SPICEN_Msk;
+
+	rx_num = 0;
+	while (rx_num < len) {
+		cnt = spic_reg->RXFLR;
+
+		for (i = 0; i < (cnt / 4); i++) {
+			fifo = spic_reg->DR.WORD;
+			memcpy((void *)(rx_data + rx_num), (void *)&fifo, 4);
+			rx_num += 4;
+		}
+
+		if (((rx_num / 4) == (len / 4)) && (rx_num < len)) {
+			for (i = 0; i < (cnt % 4); i++) {
```

Comment:
I think you can simplify by just checking if `cnt` is a multiple of 4.  You can drop the `if` check completely.


```suggestion
		for (i = 0; i < (cnt % 4); i++) {
```